app/repositories/ingame_logins.py

```diff
@@ -0,0 +1,233 @@
+    """
+    params = {
+        "id": rec_id,
+    }
+    ingame_login = await app.state.services.database.fetch_one(query, params)
+
+    assert ingame_login is not None
+    return cast(IngameLogin, ingame_login)
+
+
+async def fetch_one(
+    id: int | None = None,
+    user_id: int | None = None,
+    ip: str | None = None,
+    osu_ver: str | None = None,
+    osu_stream: str | None = None,
+) -> IngameLogin | None:
+    """Fetch a login entry from the database."""
+    if (
+        id is None
+        and user_id is None
+        and ip is None
+        and osu_ver is None
+        and osu_stream is None
+    ):
+        raise ValueError("Must provide at least one parameter.")
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM ingame_logins
+         WHERE id = COALESCE(:id, id)
+           AND userid = COALESCE(:userid, userid)
+           AND ip = COALESCE(:ip, ip)
+           AND osu_ver = COALESCE(:osu_ver, osu_ver)
+           AND osu_stream = COALESCE(:osu_stream, osu_stream)
+    """
+    params = {
+        "id": id,
+        "userid": user_id,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+    ingame_login = await app.state.services.database.fetch_one(query, params)
+
+    return cast(IngameLogin, ingame_login) if ingame_login is not None else None
+
+
+async def fetch_count(
+    user_id: int | None = None,
+    ip: str | None = None,
+) -> int:
+    """Fetch the number of logins in the database."""
+    query = """\
+        SELECT COUNT(*) AS count
+          FROM ingame_logins
+        WHERE userid = COALESCE(:userid, userid)
+          AND ip = COALESCE(:ip, ip)
+    """
+    params = {
+        "userid": user_id,
+        "ip": ip,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    assert rec is not None
+    return rec["count"]
+
+
+async def fetch_many(
+    user_id: int | None = None,
+    ip: str | None = None,
+    osu_ver: str | None = None,
+    osu_stream: str | None = None,
+    page: int | None = None,
+    page_size: int | None = None,
+) -> list[IngameLogin]:
+    """Fetch a list of logins from the database."""
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM ingame_logins
+         WHERE userid = COALESCE(:userid, userid)
+           AND ip = COALESCE(:ip, ip)
+           AND osu_ver = COALESCE(:osu_ver, osu_ver)
+           AND osu_stream = COALESCE(:osu_stream, osu_stream)
+    """
+    params = {
+        "userid": user_id,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+
+    if page is not None and page_size is not None:
+        query += """\
+            LIMIT :limit
+           OFFSET :offset
+        """
+        params["limit"] = page_size
+        params["offset"] = (page - 1) * page_size
+
+    ingame_logins = await app.state.services.database.fetch_all(query, params)
+    return cast(list[IngameLogin], ingame_logins)
+
+
+async def update(
```

Comment:
i think ingame_logins should be only creatable and not updatable - it's a historical record